lib/core/src/Cardano/Wallet/DB/Sqlite/CheckpointsOld.hs

```diff
@@ -204,23 +207,52 @@ mkStoreWalletsCheckpoints = Store{loadS=load,writeS=write,updateS=update}
     load = do
         wids <- fmap (view #walId . entityVal) <$> selectAll
         runExceptT $ do
-            xs <- forM wids $ ExceptT . loadS . mkStoreCheckpoints
+            xs <- forM wids $ ExceptT . loadS . mkStoreWallet
             pure $ Map.fromList (zip wids xs)
       where
         selectAll :: SqlPersistT IO [Entity Wallet]
         selectAll = selectList [] []
 
--- | Store for 'Checkpoints' of a single wallet.
+-- | Store for 'WalletState' of a single wallet.
+mkStoreWallet
+    :: forall s. PersistAddressBook s
+    => W.WalletId
+    -> Store (SqlPersistT IO) (DeltaWalletState s)
+mkStoreWallet wid =
```

Comment:
I wonder if we can introduce reasonable `invariant` to make sure `DeltaWalletState s` is legal/not corrupt (especially in `updateS`). It could be not the case now (as we have prologue and checkpoints only), but when new things like  tx history, pending txs, etc. will be added maybe it is going to be crucial this is the case. Do you have opinion on that?

Reply:
That's a good question: As deltas are now first-class values, how can we ensure that deltas `da` are always paired with the `Base da` value from which they were derived? 🤔

Unfortunately, I don't have a good answer to that yet, but a couple of thoughts:

1. The status quo has the same problem, so at least we are not doing worse. For example, random calls to `putTxHistory` could also mess up the database, it's the responsibility of the caller of `putTxHistory` to make sure that the call is appropriate for the current state of the database.
2. Maybe first-class deltas that can be applied to different values is something we actually want?

    * For example, the `DeltaSet1` type from `Data.Delta` satisfies interesting algebraic properties such as `apply [Insert a, Delete a] = apply (Insert a)`.
    * In this light, it is sensible to design the delta type so that applying it will never fail. For example, `Data.Set.delete` never fails to return a result, even if the element that is to be deleted is not in the set. An alternative design would have been `Data.Set.delete :: Set a -> Maybe (Set a)`, but that would have been more cumbersome to use.
    * Given a partial function `changeP :: b -> Maybe b`, we can always turn it into a total function `changeT :: b -> b` by defining `changeT b = fromMaybe b $ changeP b`. So, the idea would be to always use the second definition when designing a delta type and implementing `apply`.

3. In the other direction, perhaps it is possible to use the type system to ensure that each delta is paired with the value? In a dependently typed language, one could make the value part of the delta type. That would be very cumbersome, though. But perhaps it is possible to simulate this with some type "tags" that keep track of which delta belongs to which value? 

    ```
    newtype Delta da x y
    newtype Base da x

    apply :: Delta da x y -> Base da x -> Base da y
    ```
    Here, the tag `x` ensures that the base value and the delta belong together. Chances are that managing these tags will involve existential quantification and be somewhat cumbersome as well, but maybe there is some low-hanging fruit here.

For now, I'm leaning towards 1 and 2 — making sure that deltas can always be applied and use property tests to make that sequences of deltas are correctly generated.

Reply:
indeed, 1 and 2 sounds reasonable